src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,713 @@
+						generic_axis_get_state<int32_t>, relaxis+i);
+				}
+			}
+
+		mapi = 0;
+		for(int i=0; i<ABS_MAX; i++)
+			if(info.absaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || absaxis_map[mapi].evdev_id) && absaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(absaxis_map[mapi].evdev_id == i) {
+					if(fd >= 0)
+						ioctl(fd, EVIOCGABS(i), &absaxis[i].absinfo);
+					device()->add_item(absaxis_map[mapi].name, absaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, &absaxis[i].normalized);
+				}
+			}
+	}
+
+	void poll(void) override
+	{
+		static input_event ie[256]; // try to consume as many as possible at once
+		while(fd >= 0) {
+			ssize_t rd = read(fd, ie, 256*sizeof(input_event));
+			if(rd < 0) {
+				if(errno==EWOULDBLOCK || errno==EAGAIN)
+					return;
+				// cope with errors here
+			}
+			if(rd < sizeof(input_event))
+				return;
+			int nie = rd/sizeof(input_event);
+			for(int i=0; i<nie; i++) {
+				switch(ie[i].type) {
+					case EV_KEY:
+						keystate[ie[i].code] = (ie[i].value>0)? 0x80: 0;
+						break;
+					case EV_REL:
+						relaxis[ie[i].code] += ie[i].value;
+						break;
+					case EV_ABS: {
+						  auto& axis = absaxis[ie[i].code];
+						  axis.normalized = normalize_absolute_axis(ie[i].value, axis.absinfo.minimum, axis.absinfo.maximum);
+						}
+						break;
+				}
+			}
+		}
+	}
+
+	void reset(void) override
+	{
+		memset(keystate, 0, KEY_MAX);
+		memset(relaxis, 0, REL_MAX*sizeof(int32_t));
+		memset(absaxis, 0, ABS_MAX*sizeof(absaxis[0]));
+	}
+};
+
+#define MAX_EVDEV_DEVICES 64
+
+class evdev_input
+{
+private:
+	bool initialized;
+
+public:
+	int  devinput;
+	evdev_device_info dev[MAX_EVDEV_DEVICES]; // 640k should be enough for anyone
+	int max_dev;
+
+public:
+	evdev_input(): initialized(false)
+	{
+	}
+
+	void init(void)
+	{
+		if(initialized)
+			return;
+		initialized = true;
+		osd_printf_verbose("evdev: enumerating devices\n");
+		devinput = open("/dev/input", O_RDONLY|O_CLOEXEC|O_DIRECTORY);
+		if(devinput < 0) {
+			return;
+		}
+
+		max_dev = 0;
+		for(int i=0; i<MAX_EVDEV_DEVICES; i++) {
+			dev[i].mapped = 0;
+			dev[i].type = EVDEV_NONE;
+		}
+
+		DIR* idir = fdopendir(devinput);
+		while(idir) {
+			dirent* de = readdir(idir);
+			if(!de)
+				break;
+			if(strncmp(de->d_name, "event", 5))
+				continue;
+			int dnum = atoi(de->d_name+5);
+			if(dnum >= MAX_EVDEV_DEVICES) {
+				// too many to handle
+				continue;
+			}
+			if(dnum < MAX_EVDEV_DEVICES) {
+				if(dev[dnum].type != EVDEV_NONE) {
+					// the same device twice?
+					continue;
+				}
+			}
+			int ed = openat(devinput, de->d_name, O_RDONLY);
+			if(ed < 0) {
+				osd_printf_verbose("evdev: unable to open de->d_name: %s\n", strerror(errno));
+				continue;
+			}
+			input_id iid;
+			if(ioctl(ed, EVIOCGID, &iid)<0) {
+				osd_printf_verbose("evdev: event%d doesn't looks usable\n", dnum);
+				close(ed);
+				continue;
+			} else
+				snprintf(dev[dnum].id, 32, "%hu:%04hx:%04hx", iid.bustype, iid.vendor, iid.product);
```

Comment:
This isn’t going to produce unique IDs when someone has two similar devices connected (e.g. two Logitech Dual Action gamepads).  Can you expose the device serial number, or failing that, the datalink address for Bluetooth devices, or something derived from bus topology for USB devices?

Also, you should use `util::string_format` as it’s type-safe etc.